test/performance/range/gap_decorator_benchmark.cpp

```diff
@@ -0,0 +1,460 @@
+
+    while (gap_pos < gaps.size() && gap_acc + letter_acc < seq_len)
+    {
+        if (!gaps[gap_pos])
+            ++letter_acc;
+        else
+        {
+            if (letter_acc + gap_acc + gaps[gap_pos] > seq_len)
+            {
+                gaps[gap_pos] = seq_len - gap_acc - letter_acc;
+                gap_acc += gaps[gap_pos];
+                ++gap_pos;
+                break;
+            }
+            else
+                gap_acc += gaps[gap_pos];
+        }
+        ++gap_pos;
+    }
+    seq.resize(std::max<size_type>(1, letter_acc));  // resize ungapped sequence
+    gaps.resize(gap_pos);      // trim sampled gap vector
+}
+
+/* Helper function to prepare a gapped sequence for the benchmark (case gap_flag=true)
+ *
+ * Parameters:
+ * size_type        size type of the sequence
+ * iterator_type    iterator type of decorated sequence iterator
+ * gap_decorator_t  gap decorator type, e.g. gap_decorator_anchor_set
+ *
+ * Arguments:
+ * gaps             reference to gap vector
+ * gap_decorator    reference to gap decorator
+ */
+template<typename size_type, typename iterator_type, typename gap_decorator_t>
+void insert_gaps(std::vector<size_type> & gaps, gap_decorator_t & gap_decorator)
+{
+    [[maybe_unused]] size_type gap_acc = 0;
+    for (size_type i = 0; i < gaps.size(); ++i)
+    {
+        if (gaps[i])
+        {
+            iterator_type it = std::ranges::begin(gap_decorator);
+            std::ranges::advance(it, std::min(i + gap_acc, gap_decorator.size()));
+            insert_gap(gap_decorator, it, gaps[i]);
+        }
+        gap_acc += gaps[i];
+    }
+}
+
+// ============================================================================
+//  read left to right (looped in case #ops exceeds sequence length)
+// ============================================================================
+/* Parameters:
+ * gap_decorator_t      gap decorator class, e.g. gap_decorator_anchor_set
+ * unaligned_seq_t     unaligned sequence to which a reference is stored
+ * alphabet_t           alphabet type of the ungapped sequence
+ * gapped_flag          operate on already gapped (true) or ungapped sequence (false)
+ */
+
+
+template <typename gap_decorator_t, bool gapped_flag>
+static void read_left2right(benchmark::State& state)
+{
+    // get target sequence length from current range state
+    unsigned int seq_len = state.range(0);
+    using size_type = typename gap_decorator_t::size_type;
+    using iterator_type = typename gap_decorator_t::iterator;
+    using sequence_type = remove_cvref_t<detail::unaligned_seq_t<gap_decorator_t>>;
+    sequence_type seq(seq_len, 'A'_dna4);
+
+    // vector of sampled gap lengths for each position
+    std::vector<size_type> gaps(seq_len, 0);
+    sample<size_type>(&gaps, seq_len);
+
+    // determine sum of gaps and non-gap symbols for not exceeding targeted sequence length
+    if constexpr(gapped_flag)
+        resize<size_type, sequence_type>(gaps, seq, seq_len);
```

Comment:
Because first the aligned sequence length (some power of 2) is fixed, and then I sample the gap lengths. Hence, I cannot know in advance how long the ungapped sequence has to be.

Reply:
Do you need the gapped length explicitly? Otherwise, it seems easier to fix the ungapped length and then just insert gaps in the aligned sequence.

Reply:
But it is working and is not a performance issue. It is clearer for benchmarking to fix the aligned sequence length compared to fixing only the aligned sequence length and not giving any guarantees how long the actual sequence length will be (due to gap sampling). It makes a huge difference if there are only 1% gaps (relative to unaligned sequence length) inserted or 90%. If I obfuscate this, the interpretation and comparison of the runtime results would become harder.

Reply:
Initially, I designed the benchmark to have a final gap proportion of roughly 40-45 %. For each position gap lengths are sampled according to a distribution that I found in a paper where they looked at  a typical use case and counted the gap lengths. I wrote this when posting the first benchmark results. But I agree, that we should be flexible here.
Comparing few long gaps vs. many small gaps is another issue. That would mean having now two distributions that are being sampled and would open again another parameter dimension in the performance benchmark. 
I doubt that those results give essential insights for the following reason: it is improbable that we ever use a bit vector version (due to bad performance in previous benchmarks), hence, the current approaches store gaps as anchors, so they always manage same sized data sets. Though, they have different theoretical runtimes for random access and those differences show up through simple scaling of the target length - double length means double amount of gaps. 
So, we measure few versus many gaps, but do not vary gap lengths. The latter is something to consider when using approaches storing gaps as single symbols, e.g. with bit vectors.
Varying gap density shifts the crossing points (where A outperforms B) on the x-axis, e.g. the anchor block approach outperforms the anchor set approach already at half the sequence length compared to a run with halvened gap density.
If everyone agrees I would then add the gap density parameter. In this regard we should think about some plotting functionality or csv output, s.t. results can be transformed by e.g. a tabular program to better investigate the performance.

Reply:
One last remark: due to sampling of gap lengths I cannot now in advance the ungapped sequence length if a fixed aligned sequence length is targeted. Even if I now will aim for 5 %, 10 %, ... relative gap amount, it is still sampling involved. Insertion would then stop, when roughly 5 %, 10 %, etc. are reached. So, I recommend to keep the resizing of the underlying sequence, but introduce varying gap densities. Limiting the random behaviour also reduces the expressiveness of the benchmark.